### ISSUE TITLE
Moq1203 false negative: name-based fallback fires even when the chained method resolved to a non-Moq symbol

### ISSUE THREAD

**Issue by rjmurillo** [rjmurillo/moq.analyzers#1243]
## Summary

An internal code audit found that `MethodSetupShouldSpecifyReturnValueAnalyzer` (rule **Moq1203** — note: the audit draft mislabeled this as Moq1201; Moq1201 is `AsyncUsesReturnsAsyncInsteadOfResult` per `src/Common/DiagnosticIds.cs:15,17`) applies its string-name fallback unconditionally: any method in the fluent chain *named* `Returns`, `ReturnsAsync`, `Throws`, or `ThrowsAsync` suppresses the diagnostic, even when Roslyn successfully resolved the call to a **non-Moq user method**. This is a name-based semantic decision that ADR-001 forbids and produces false negatives. **Severity: Medium.** The fix must not regress the case the fallback was added for: Moq's delegate-based `ReturnsAsync` overloads live in `Moq.GeneratedReturnsExtensions` (verified present in both Moq 4.8.2 and 4.18.4), a class `MoqKnownSymbols` does not track, so those calls resolve successfully but fail `IsInstanceOf` — the proper fix is to track that class as a known symbol *and* gate the name fallback on failed resolution.

## Current behavior

`src/Analyzers/MethodSetupShouldSpecifyReturnValueAnalyzer.cs:179-189` (inside the `HasReturnValueSpecification` chain walk):

```csharp
            // First try semantic symbol matching (exact). Then fall back to method
            // name matching when Roslyn resolves a symbol that IsInstanceOf cannot
            // match (e.g., delegate-based overloads with constructed generic types).
            // The name-based fallback is safe because this walk only visits methods
            // chained from a verified Setup() call, and we verify the named method
            // exists in the Moq compilation.
            if (HasReturnValueSymbol(symbolInfo, knownSymbols)
                || IsKnownReturnValueMethodName(memberAccess.Name.Identifier.ValueText, knownSymbols))
            {
                return true;
            }
```

`IsKnownReturnValueMethodName` (`src/Analyzers/MethodSetupShouldSpecifyReturnValueAnalyzer.cs:211-223`) only checks that Moq's assembly has *some* method with that name — not that this call resolves to it:

```csharp
    private static bool IsKnownReturnValueMethodName(string methodName, MoqKnownSymbols knownSymbols)
    {
        return methodName switch
        {
            "Returns" => !knownSymbols.IReturnsReturns.IsEmpty
                         || !knownSymbols.IReturns1Returns.IsEmpty
                         || !knownSymbols.IReturns2Returns.IsEmpty,
            "ReturnsAsync" => !knownSymbols.ReturnsExtensionsReturnsAsync.IsEmpty,
            "Throws" => !knownSymbols.IThrowsThrows.IsEmpty,
            "ThrowsAsync" => !knownSymbols.ReturnsExtensionsThrowsAsync.IsEmpty,
            _ => false,
        };
    }
```

`HasReturnValueSymbol` (`src/Analyzers/MethodSetupShouldSpecifyReturnValueAnalyzer.cs:230-240`) already handles both the resolved-symbol and the candidate-symbols (overload-resolution-failure) cases:

```csharp
    private static bool HasReturnValueSymbol(SymbolInfo symbolInfo, MoqKnownSymbols knownSymbols)
    {
        if (symbolInfo.Symbol is IMethodSymbol resolved)
        {
            return resolved.IsMoqReturnValueSpecificationMethod(knownSymbols);
        }

        return symbolInfo.CandidateSymbols
            .OfType<IMethodSymbol>()
            .Any(method => method.IsMoqReturnValueSpecificationMethod(knownSymbols));
    }
```

**False-negative scenario.** A user-defined fluent helper that merely shares a Moq name suppresses Moq1203 purely by identifier text:

```csharp
public static class UserFluentExtensions
{
    // Logs; does NOT configure a return value, returns void.
    public static void Returns<TMock, TResult>(this Moq.Language.Flow.ISetup<TMock, TResult> setup, string description)
        where TMock : class { }
}

var mock = new Mock<IFoo>();
mock.Setup(x => x.GetValue()).Returns("log only");   // resolves to the USER method
// Expected: Moq1203 (no return value specified). Actual: no diagnostic (name fallback).
```

`HasReturnValueSymbol` correctly returns false (resolved symbol is not a Moq method), and `IsFluentChainWrapperMethod` (lines 195-198) correctly returns false (return type `void` is not a Moq fluent type) — but `IsKnownReturnValueMethodName("Returns", ...)` short-circuits the walk to "handled".

**Why the fallback exists (root cause, verified against the shipped Moq assemblies).** Moq hosts its delegate-based `ReturnsAsync` overloads (`ReturnsAsync<TMock, ..., TResult>(this IReturns<TMock, Task<TResult>>, Func<...>)`, arities `Func<T1,TResult>` through `Func<T1..T15,TResult>`) in a **separate static class `Moq.GeneratedReturnsExtensions`** — in *both* Moq 4.8.2 and 4.18.4 — while `MoqKnownSymbols` only tracks `Moq.ReturnsExtensions` (`src/Common/WellKnown/MoqKnownSymbols.cs:348`). So for the existing test row (`tests/Moq.Analyzers.Test/MethodSetupShouldSpecifyReturnValueAnalyzerTests.cs`, `CustomReturnTypeTestData`, "Delegate-based ReturnsAsync" rows):

```csharp
databaseMock.Setup(mock => mock.SaveAsync(It.IsAny<MyValue>())).ReturnsAsync((MyValue val) => val);
```

resolution **succeeds** (Symbol = reduced `GeneratedReturnsExtensions.ReturnsAsync`), `IsInstanceOf(knownSymbols.ReturnsExtensionsReturnsAsync)` fails, and today only the name fallback (and, since #1086, incidentally the fluent-wrapper heuristic — `GeneratedReturnsExtensions.ReturnsAsync` returns `Moq.Language.Flow.IReturnsResult<TMock>`) keeps the row green. A naive "gate on `Symbol is null`" fix without adding the known symbol would make this case depend solely on the wrapper heuristic, which is fragile; the principled ADR-001/ADR-006 fix is to track `GeneratedReturnsExtensions` explicitly.

## Desired behavior

- A chained method that **resolved** (or produced candidate symbols) is judged only by symbols: genuine Moq return-value methods (including `GeneratedReturnsExtensions.ReturnsAsync`) suppress Moq1203; non-Moq methods do not (unless the fluent-wrapper heuristic of #1086 applies, which is intentional and unchanged).
- The string-name fallback applies **only** when Roslyn produced no symbol information at all (`Symbol is null` and `CandidateSymbols.IsEmpty`) — i.e. genuinely unresolvable mid-edit code, where conservative suppression avoids false positives.

## Implementation plan

1. **`src/Common/WellKnown/MoqKnownSymbols.cs`** — track `Moq.GeneratedReturnsExtensions`:

   a. Add a field after the existing `_returnsExtensionsThrowsAsync` field (line 33):

```csharp
    private readonly Lazy<ImmutableArray<IMethodSymbol>> _generatedReturnsExtensionsReturnsAsync;
```

   b. Initialize it in the constructor, immediately after line 113 (`_returnsExtensionsThrowsAsync = CreateLazyMethods(ReturnsExtensions, "ThrowsAsync");`):

```csharp
        _generatedReturnsExtensionsReturnsAsync = CreateLazyMethods(GeneratedReturnsExtensions, "ReturnsAsync");
```

   c. Add the type and methods properties next to the existing `ReturnsExtensions*` properties (after line 358):

```csharp
    /// <summary>
    /// Gets the class <c>Moq.GeneratedReturnsExtensions</c>, which hosts Moq's
    /// generated delegate-based <c>ReturnsAsync</c> overloads.
    /// </summary>
    internal INamedTypeSymbol? GeneratedReturnsExtensions => TypeProvider.GetOrCreateTypeByMetadataName("Moq.GeneratedReturnsExtensions");

    /// <summary>
    /// Gets the methods for <c>Moq.GeneratedReturnsExtensions.ReturnsAsync</c>.
    /// </summary>
    internal ImmutableArray<IMethodSymbol> GeneratedReturnsExtensionsReturnsAsync => _generatedReturnsExtensionsReturnsAsync.Value;
```

   (`CreateLazyMethods` already accepts a nullable type and yields an empty array when the type is absent, so no null handling is needed. `GeneratedReturnsExtensions` has no `ThrowsAsync` members in either supported Moq version — only `ReturnsAsync` — so no ThrowsAsync counterpart is added.)

2. **`src/Common/ISymbolExtensions.Moq.cs`** — replace `IsMoqReturnsAsyncMethod` (lines 97-106) with:

```csharp
    /// <summary>
    /// Determines whether a symbol is a Moq ReturnsAsync extension method.
    /// </summary>
    /// <param name="symbol">The symbol to check.</param>
    /// <param name="knownSymbols">The known symbols for type checking.</param>
    /// <returns>True if the symbol is a ReturnsAsync method from Moq.ReturnsExtensions or Moq.GeneratedReturnsExtensions; otherwise false.</returns>
    internal static bool IsMoqReturnsAsyncMethod(this ISymbol symbol, MoqKnownSymbols knownSymbols)
    {
        return symbol.IsInstanceOf(knownSymbols.ReturnsExtensionsReturnsAsync)
            || symbol.IsInstanceOf(knownSymbols.GeneratedReturnsExtensionsReturnsAsync);
    }
```

   (`IsMoqReturnsAsyncMethod` is consumed only via `IsMoqReturnValueSpecificationMethod`, which is consumed only by this analyzer — grep-verified — so this widening has no side effects on other rules.)

3. **`src/Analyzers/MethodSetupShouldSpecifyReturnValueAnalyzer.cs`** — gate the fallback. Replace the fallback block at lines 179-189 with:

```csharp
            // First try semantic symbol matching (exact), including candidate symbols
            // when overload resolution fails. Fall back to method-name matching ONLY
            // when Roslyn produced no symbol information at all (unresolvable mid-edit
            // code); we then verify the named method exists in the referenced Moq
            // assembly. Applying the name fallback when a symbol resolved would let
            // user-defined methods that merely share a Moq method name (Returns,
            // ReturnsAsync, Throws, ThrowsAsync) suppress the diagnostic (ADR-001).
            bool resolutionProducedNoSymbols = symbolInfo.Symbol is null && symbolInfo.CandidateSymbols.IsEmpty;

            if (HasReturnValueSymbol(symbolInfo, knownSymbols)
                || (resolutionProducedNoSymbols && IsKnownReturnValueMethodName(memberAccess.Name.Identifier.ValueText, knownSymbols)))
            {
                return true;
            }
```

   And replace the `"ReturnsAsync"` arm of `IsKnownReturnValueMethodName` (line 218) with:

```csharp
            "ReturnsAsync" => !knownSymbols.ReturnsExtensionsReturnsAsync.IsEmpty
                              || !knownSymbols.GeneratedReturnsExtensionsReturnsAsync.IsEmpty,
```

   Also update the doc comment above `IsKnownReturnValueMethodName` (lines 206-210) — its "last-resort fallback when Roslyn cannot resolve symbols at all" wording is now literally enforced; no wording change strictly required, but keep it consistent if touched.

4. **Why every existing test still passes under the new predicate** (walked file `tests/Moq.Analyzers.Test/MethodSetupShouldSpecifyReturnValueAnalyzerTests.cs`):
   - `TestData`, `SemanticVariationSetupTestData`, `Issue887_*`: `.Returns/.Throws/.ReturnsAsync/.ThrowsAsync` resolve to genuine Moq symbols → `HasReturnValueSymbol` matches (unchanged).
   - `OverloadResolutionFailureTestData` (wrong argument types): `SymbolInfo.CandidateSymbols` is non-empty and contains Moq methods → matched by the candidate scan in `HasReturnValueSymbol`; the fallback gate (`CandidateSymbols.IsEmpty`) is irrelevant.
   - `OverloadResolutionFailureWithDiagnosticTestData` (`.Callback("wrong type")` with diagnostic): candidates are Callback methods, name `Callback` is not in the fallback list → still flags.
   - `CustomReturnTypeTestData` delegate-based `ReturnsAsync` rows (the fallback's original motivation): now matched *directly* by `HasReturnValueSymbol` via `GeneratedReturnsExtensionsReturnsAsync` (step 1-2) on both Moq 4.8.2 and 4.18.4 — no longer dependent on the name fallback or the wrapper heuristic.
   - `Issue849_*`, `CallbackOnlyNewMoqTestData`: Callback is excluded from the wrapper heuristic and never matched the name list; walks continue exactly as before.
   - `Issue1067_*`: user extensions have non-Moq names (`ReturnsTrue`, `LogSetup`, ...) — the name fallback never applied to them; wrapper heuristic behavior unchanged.

5. **Docs**: update `docs/rules/Moq1203.md` with a short note that user-defined chain methods named like Moq's return-value methods do **not** suppress the diagnostic unless they return a Moq fluent type (wrapper pattern, see issue #1067 behavior).

## Tests

**File: `tests/Moq.Analyzers.Test/MethodSetupShouldSpecifyReturnValueAnalyzerTests.cs`** — add a data source, a source builder, and a theory:

```csharp
    // Regression data for the name-based fallback overreach: user-defined extension
    // methods that merely share a Moq return-value method name must NOT suppress
    // Moq1203, unless they return a Moq fluent type (wrapper pattern, #1067).
    public static IEnumerable<object[]> UserMethodsNamedLikeMoqReturnValueMethodsTestData()
    {
        IEnumerable<object[]> data =
        [

            // Positive: user extension named Returns (returns void) — diagnostic expected
            ["""
            var mock = new Mock<IFoo>();
            {|Moq1203:mock.Setup(x => x.GetValue())|}.Returns("log only");
            """],

            // Positive: user extension named ReturnsAsync (returns string) — diagnostic expected
            ["""
            var mock = new Mock<IFoo>();
            {|Moq1203:mock.Setup(x => x.BarAsync())|}.ReturnsAsync("log only");
            """],

            // Positive: user Returns chained after a real Callback — diagnostic expected
            ["""
            var mock = new Mock<IFoo>();
            {|Moq1203:mock.Setup(x => x.GetValue())|}.Callback(() => { }).Returns("log only");
            """],

            // Negative: user extension named Returns that RETURNS a Moq fluent type is
            // treated as a wrapper (#1067 behavior preserved) — no diagnostic
            ["""
            var mock = new Mock<IFoo>();
            mock.Setup(x => x.GetValue()).Returns(42, "described");
            """],
        ];

        return data.WithNamespaces().WithNewMoqReferenceAssemblyGroups();
    }

    [Theory]
    [MemberData(nameof(UserMethodsNamedLikeMoqReturnValueMethodsTestData))]
    public async Task ShouldFlagSetupWhenUserMethodMerelySharesMoqReturnValueName(string referenceAssemblyGroup, string @namespace, string mock)
    {
        string source = BuildUserNamedExtensionSource(@namespace, mock);
        output.WriteLine(source);

        await Verifier.VerifyAnalyzerAsync(source, referenceAssemblyGroup).ConfigureAwait(false);
    }

    private static string BuildUserNamedExtensionSource(string @namespace, string mock)
    {
        return $$"""
            {{@namespace}}
            using Moq.Language.Flow;

            public interface IFoo
            {
                int GetValue();
                Task<int> BarAsync();
            }

            public static class UserFluentExtensions
            {
                // Same names as Moq's return-value methods, but user-defined and
                // NOT configuring any return value.
                public static void Returns<TMock, TResult>(this ISetup<TMock, TResult> setup, string description)
                    where TMock : class
                {
                }

                public static void Returns<TMock, TResult>(this IReturnsThrows<TMock, TResult> setup, string description)
                    where TMock : class
                {
                }

                public static string ReturnsAsync<TMock, TResult>(this ISetup<TMock, TResult> setup, string description)
                    where TMock : class
                {
                    return description;
                }

                // Returns a Moq fluent type: intentionally still suppresses Moq1203 (wrapper pattern).
                public static ISetup<TMock, int> Returns<TMock>(this ISetup<TMock, int> setup, int value, string description)
                    where TMock : class
                {
                    setup.Returns(value);
                    return setup;
                }
            }

            internal class UnitTest
            {
                private void Test()
                {
                    {{mock}}
                }
            }
            """;
    }
```

Notes: the builder mirrors `BuildWrappedSetupSource` in the same file (file-scoped-namespace-then-usings layout is already proven there); `WithNewMoqReferenceAssemblyGroups()` matches the Issue1067 suites that also declare Moq fluent-interface extensions. Resolution check for the positive rows: Moq's instance `Returns(int)` / `Returns(Func<int>)` overloads are inapplicable to a `string` argument, so the user extension binds and `SymbolInfo.Symbol` is the (reduced) user method — exactly the overreach case.

**File: `tests/Moq.Analyzers.Test/Common/MoqKnownSymbolsTests.ReturnsAndThrows.cs`** — add symbol-catalog tests mirroring the existing `ReturnsExtensions*` tests in that file:

```csharp
    [Fact]
    public void GeneratedReturnsExtensions_WithoutMoqReference_ReturnsNull()
    {
        MoqKnownSymbols symbols = CreateSymbolsWithoutMoq();
        Assert.Null(symbols.GeneratedReturnsExtensions);
    }

    [Fact]
    public void GeneratedReturnsExtensionsReturnsAsync_WithoutMoqReference_ReturnsEmpty()
    {
        MoqKnownSymbols symbols = CreateSymbolsWithoutMoq();
        Assert.True(symbols.GeneratedReturnsExtensionsReturnsAsync.IsEmpty);
    }

    [Fact]
    public async Task GeneratedReturnsExtensions_WithMoqReference_ReturnsNamedTypeSymbol()
    {
        MoqKnownSymbols symbols = await CreateSymbolsWithMoqAsync();
        Assert.NotNull(symbols.GeneratedReturnsExtensions);
        Assert.Equal("GeneratedReturnsExtensions", symbols.GeneratedReturnsExtensions!.Name);
    }

    [Fact]
    public async Task GeneratedReturnsExtensionsReturnsAsync_WithMoqReference_ReturnsNonEmpty()
    {
        MoqKnownSymbols symbols = await CreateSymbolsWithMoqAsync();
        Assert.False(symbols.GeneratedReturnsExtensionsReturnsAsync.IsEmpty);
    }
```

**Existing tests that must keep passing (regression gates):** the entire `MethodSetupShouldSpecifyReturnValueAnalyzerTests` class — in particular `CustomReturnTypeTestData`'s two "Delegate-based ReturnsAsync" rows (both Moq versions ×2 namespaces), `OverloadResolutionFailureTestData`, `OverloadResolutionFailureWithDiagnosticTestData`, `Issue1067_WrappedSetupTestData` (including the documented `NoOpPassthrough` accepted-false-negative row), and `Issue1067_NonFluentExtensionTestData`.

## Constraints

- Analyzer assemblies target netstandard2.0; Roslyn SDK pinned to Microsoft.CodeAnalysis **4.8** (ADR-003) — only `SymbolInfo`, `ImmutableArray`, `GetOrCreateTypeByMetadataName` are used, all available in 4.8. No new package dependencies (ADR-004/ADR-005).
- ADR-001: this change *strengthens* symbol-based detection; the residual name fallback is restricted to the no-symbol-information case and stays documented as a mid-edit conservative measure.
- ADR-006: the new symbols go through `MoqKnownSymbols` `Lazy` caching exactly like the existing entries (one `MoqKnownSymbols` per compilation; no `GetTypeByMetadataName` in per-operation callbacks).
- Tests follow repo patterns: `{|Moq1203:...|}` markup asserting ID + span; `[Theory]` + `[MemberData]` with `public static IEnumerable<object[]>`; rows multiplied via `TestDataExtensions` (new-Moq group here, consistent with the Issue1067 suites; the delegate-based regression rows already run on both 4.8.2 and 4.18.4).
- Update `docs/rules/Moq1203.md` (behavior change: fewer false negatives). Do NOT touch `AnalyzerReleases.Shipped.md`; no `AnalyzerReleases.Unshipped.md` change (no severity/category change).

## Validation

```bash
dotnet format --verify-no-changes   # or `dotnet format` before committing
dotnet build /p:PedanticMode=true   # must be 0 warnings, 0 errors
dotnet test --settings ./build/targets/tests/test.runsettings
```

## Acceptance criteria

- [ ] `MoqKnownSymbols` exposes `GeneratedReturnsExtensions` and `GeneratedReturnsExtensionsReturnsAsync` (lazy-cached, ADR-006 pattern).
- [ ] `IsMoqReturnsAsyncMethod` matches `ReturnsAsync` from both `Moq.ReturnsExtensions` and `Moq.GeneratedReturnsExtensions`.
- [ ] The name fallback in `HasReturnValueSpecification` fires only when `symbolInfo.Symbol is null && symbolInfo.CandidateSymbols.IsEmpty`.
- [ ] New positive rows report Moq1203 for user extensions named `Returns`/`ReturnsAsync` that do not return a Moq fluent type; the fluent-returning `Returns` wrapper row stays silent.
- [ ] `CustomReturnTypeTestData` delegate-based `ReturnsAsync` rows pass on **both** Moq 4.8.2 and 4.18.4 (now via direct symbol match).
- [ ] All other existing `MethodSetupShouldSpecifyReturnValueAnalyzerTests` suites pass unchanged.
- [ ] `docs/rules/Moq1203.md` updated; build 0 warnings including `/p:PedanticMode=true`.

## Out of scope

- The wrapper heuristic itself (`IsFluentChainWrapperMethod`, #1086) and its documented accepted false negative (`NoOpPassthrough`).
- Consolidating `TryGetMockedMethodSymbol` with `MoqVerificationHelpers` (audit item E-6).
- Any change to Moq1201 (`AsyncUsesReturnsAsyncInsteadOfResult`) — not affected despite the audit draft's label.
